### PR TITLE
fix: log cause property from the error

### PR DESCRIPTION
**Pull request by thetutlage**
Fixes: FRMW-2762

This PR fixes the error reporting to include the original error message and the error cause when logging it using the JSON formatter of Winston.

> Side note: There is still a lot of room for improvements in our logging. But that require a overhaul cleanup in multiple places and will require dedicated loggers for commands and APP and special treatment for pretty printing errors.

## Error output in JSON format

**Before**
CleanShot 2024-11-06 at 13 12 32@2x

**After**
CleanShot 2024-11-06 at 13 11 28@2x

### CODE REVIEW

File: packages/cli/medusa-cli/src/reporter/index.ts [hunk 1]
```diff
@@ -174,27 +175,57 @@ export class Reporter {
    * Level 0
    */
   error(messageOrError: string | Error, error?: Error) {
-    let message = messageOrError as string
+    let message: string
+    let errorAsObject: Error | undefined
 
-    if (typeof messageOrError === "object") {
+    if (messageOrError && typeof messageOrError === "object") {
```
Comment (adrien2p) [inline packages/cli/medusa-cli/src/reporter/index.ts:181]: suggestion: `isObject(messageOrError)`

File: packages/cli/medusa-cli/src/reporter/index.ts [hunk 2]
```diff
@@ -174,27 +175,57 @@ export class Reporter {
...
 
+    /**
+     * If "errorAsObject" has a message property, then we will
+     * print the standalone message as one log item and then
+     * the actual error object as another log item.
+     *
+     * Otherwise, we always loose the message property from the
+     * actual error object
+     */
+    if (errorAsObject?.message && errorAsObject?.message !== message) {
```
Comment (adrien2p) [inline packages/cli/medusa-cli/src/reporter/index.ts:217]:
```suggestion
    if (errorAsObject?.message !== message) {
```

File: packages/cli/medusa-cli/src/reporter/index.ts [hunk 3]
```diff
@@ -174,27 +175,57 @@ export class Reporter {
...
-    // Give stack traces and details in dev
-    if (error && IS_DEV) {
-      console.log(error)
+    /**
+     * In dev we print the error using `console.error`. I am not sure if
+     * we should be doing this, because now we are giving more info
+     * in Dev and less info in production
+     */
+    if (errorAsObject && IS_DEV) {
+      console.error(errorAsObject)
```
Comment (adrien2p) [inline packages/cli/medusa-cli/src/reporter/index.ts:227]: discussion: I am not sure about that either, also why are we suddenly using the console API there? I think it should be similar no matter the env right?

**Review by adrien2p — APPROVED** [review]

File: packages/cli/medusa-cli/src/reporter/index.ts (hunk 3 above)
Comment (thetutlage) [reply packages/cli/medusa-cli/src/reporter/index.ts:227]: Yeah. But I noticed that Winston doesn't display the error stack when not using the `JSON` formatter. 

This is how it looks when I remove the `console.error` statement

CleanShot 2024-11-06 at 13 59 31@2x

Comment (adrien2p) [reply packages/cli/medusa-cli/src/reporter/index.ts:227]: weird 😂 also, winston added support for cause lately 🤔

**Review by adrien2p — APPROVED** [review]

File: packages/cli/medusa-cli/src/reporter/index.ts (hunk 3 above)
Comment (thetutlage) [reply packages/cli/medusa-cli/src/reporter/index.ts:227]: They use `logform` for formatting and as per this source, the `CLI` formatter does not print anything other than the error message. https://github.com/winstonjs/logform/blob/master/cli.js

We use `CLI` formatter during development

**Pull request merged**